### ISSUE TITLE
Process Freezing and large thread count

### ISSUE THREAD

**Issue by KateWilkins** [mperham/sidekiq#3512]
Ruby version: 2.3.1p112
Sidekiq version: 4.2.1

  sidekiq_options queue: :entities, retry: false
  sidekiq_throttle({ concurrency: { limit: 1 }})

We have scanning workers that run every X seconds with sidekiq, one for each customer, there can be multiple scanner tasks and we limit the concurrency to 1 and the task requeues itself just before it completes (in the ensure a of begin/rescue/ensure wrap of the job) to run X seconds after it finishes running so that we never over-queue as we don't know how long the mining operation will take (mining realtime retail transaction data), we start the system by queuing one for each customer.

We have the God gem running above to force restarts and thats working just fine on the occasion we get an OOM problem.

The issue we have is that every week or so one of the systems will just hang, the process is still there so God hasn't stepped in to restart it and it shows as sleeping, with scheduled jobs but nothing runs. Kill the main thread and God steps in, restarts sidekiq and all is good for another week or so.

I've been reading the trouble shooting and will try the TTIN signal next time it happens but thought in the meantime I'd post the issue/question to see if anyone has a similar experience or if we're doing something wrong....

With todays fail I took a look in /proc/<pid>/status and it was showing >28000 tasks in /proc/<<pid>>/tasks/... and looking in /proc/<<pid>>/tasks/<tid>/status they're all sleeping (the 10 or so random ones I looked at). I'm not an expert on reading/inferring the meaning of the proc status data, currently I'm assuming those are 28000+ live tasks and we're just continuously spawning tasks and they're not closing for some reason and eventually we are hitting some upper task limit and hence the sleep.

Worker/Rescheduling code is below:

```
def perform(retailer_id)
  begin
    config = ActiveRecord::Base.configurations[Rails.env] ||
        Rails.application.config.database_configuration[Rails.env]
    config['reaping_frequency'] = ENV['DB_REAP_FREQ'] || 10 # seconds
    config['pool']            =   ENV['DB_POOL'] || 10
    ActiveRecord::Base.establish_connection(config)

    # mine transaction data

  rescue => e
    # Log/email failure

  ensure
      TillsWorker.requeue(retailer_id.to_i, true)
      # clear any other redundant connections (to ensure we don't run out!)
      ActiveRecord::Base.clear_active_connections!

  end
end

def self.requeue(retailer_id, force = false)
    return false if !force && has_job_scheduled?(retailer_id) # Checks scheduled/queued/running for klass/retailer_id

    # Stop any dups
    r = Sidekiq::ScheduledSet.new
    jobs = r.select { |retri| retri.klass == self.name && retri.args == [retailer_id] }
    jobs.each(&:delete)

    # ensure no queue dupes
    queue = Sidekiq::Queue.new('entities')
    queue.each do |job|
      job.delete if job.klass == self.name && job.args == [retailer_id]
    end

    # find config, if it exists, requeue
    entity_manager_config = $retailerConfig[retailer_id.to_s]

    TillsWorker.perform_in(entity_manager_config['frequency'].to_i.seconds, retailer_id)
end

```

**Comment by mperham** [comment]
1. I don't understand the purpose of the AR code?
2. Your `requeue` method will be very slow if you have lots of scheduled or enqueued jobs.  I would recommend using a cron task + unique plugin to enqueue jobs for each retailer.

Sidekiq is not designed for tasks that attempt to loop themselves.  Sidekiq Enterprise has cron+unique features or you can find OSS plugins that might work.

**Comment by KateWilkins** [comment]
Re the questions:

1. I should have removed when I pruned the other code to past into the issue, however it looks like it's probably is the cause of the problem as when I disable it then the thread count doesn't increase each time the worker runs. It seems it was added due to issues with running out of DB pool connections. I've moved it to application startup instead and removed it from the worker.

2. We only ever have about 20 max jobs in the scheduled queue so its not so much of an issue.

Thanks for the comment, I will close and come back if it persists.
